README.md
<p align="center">
  <img src="/animation.gif" />
</p>

## Installation
`MessageViewController` is available through CocoaPods and Carthage. Just add it to your Podfile or Cartfile and install. Done!
### CocoaPods
```ruby
pod 'MessageViewController'
```
### Carthage
```
github "GitHawkApp/MessageViewController"
```

## Setup

You must subclass `MessageViewController`.

```swift
import MessageViewController

class ViewController: MessageViewController {
  // ...
}
```

Finish setup using a `UIScrollView`. Remember this can also be a `UITableView` or `UICollectionView`.

```swift
func viewDidLoad() {
  super.viewDidLoad()
  setup(scrollView: scrollView)
}
```

## Customizations

You can customize any part of the UI that you want!

```swift
// Border between the text view and the scroll view
borderColor = .lightGray

// Change the appearance of the text view and its content
messageView.inset = UIEdgeInsets(top: 8, left: 16, bottom: 8, right: 16)
messageView.textView.placeholderText = "New message..."
messageView.textView.placeholderTextColor = .lightGray
messageView.font = .systemFont(ofSize: 17)

// Setup the button using text or an icon
messageView.set(buttonTitle: "Send", for: .normal)
messageView.addButton(target: self, action: #selector(onButton))
messageView.buttonTint = .blue

// Set custom attributes for an autocompleted string
let tintColor = .blue
messageAutocompleteController.autocompleteTextAttributes = ["@": [.font: UIFont.preferredFont(forTextStyle: .body), .foregroundColor: tintColor, .backgroundColor: tintColor.withAlphaComponent(0.1)]]
```

## Autocomplete

The base view controller uses a `MessageAutocompleteController` control to handle text autocompletion.

This control uses a plain `UITableView` to display its autocomplete. Add a `dataSource` and `delegate` to display and handle interactions.

```swift
let tableView = messageAutocompleteController.tableView
tableView.register(UITableViewCell.self, forCellReuseIdentifier: "cell")
tableView.dataSource = self
tableView.delegate = self
```

Then register for autocomplete prefixes you want to respond to and set a `delegate` to handle when a prefix is found.

```swift
messageAutocompleteController.register(prefix: "@")
messageAutocompleteController.delegate = self
```

Your delegate needs to implement just one method.

```swift
func didFind(controller: MessageAutocompleteController, prefix: String, word: String) {
  // filter your data
  controller.show(true)
}
```

> **Note:** You can perform asyncronous autocomplete searches. Just be sure to call `messageAutocompleteController.show()` when finished.

## Acknowledgements

- Heavy inspiration from [SlackTextViewController](https://github.com/slackhq/SlackTextViewController)
- Created with ❤️ by [Ryan Nystrom](https://twitter.com/_ryannystrom)


CHANGELOG.md
# Next

## Added

- **Breaking Change:** Requires Xcode 10.2+

- Upgraded to Swift 5 and added Swift 5 support to the podspec. [#91](https://github.com/GitHawkApp/MessageViewController/pull/91) by [@benasher44](https://github.com/benasher44)

# 0.2.0

## Added

- **Breaking Change:** Migrated to Swift 4.2

- **Breaking Change:** Added a new delegate method to `MessageTextViewListener`, `func willChangeRange(textView: MessageTextView, to range: NSRange)` which allowed for the observation of text range changes such that the entire autocomplete string is deleted rather than character by character. [#15](https://github.com/GitHawkApp/MessageViewController/pull/15) by [@nathantannar4](https://github.com/nathantannar4)

- Added an optional `accessibilityLabel` parameter to `setButton`, and using the supplied title or image label by default. [#57](https://github.com/GitHawkApp/MessageViewController/pull/57) by [@BasThomas](https://github.com/BasThomas)

## Removed

- Removed a tap gesture recognizer on the message view that would call `becomeFirstResponder()` on the text view. [#9](https://github.com/GitHawkApp/MessageViewController/pull/9) by [@rizwankce](https://github.com/rizwankce)

## Fixed

## Miscellaneous

# 0.1.1

Initial release


Examples/Examples/ViewController.swift
//
//  ViewController.swift
//  Examples
//
//  Created by Ryan Nystrom on 1/28/18.
//  Copyright © 2018 Ryan Nystrom. All rights reserved.
//

import UIKit
import MessageViewController

class ViewController: MessageViewController, UITableViewDataSource, UITableViewDelegate, MessageAutocompleteControllerDelegate {

    var data = "Lorem ipsum dolor sit amet|consectetur adipiscing elit|sed do eiusmod|tempor incididunt|ut labore et dolore|magna aliqua| Ut enim ad minim|veniam, quis nostrud|exercitation ullamco|laboris nisi ut aliquip|ex ea commodo consequat|Duis aute|irure dolor in reprehenderit|in voluptate|velit esse cillum|dolore eu|fugiat nulla pariatur|Excepteur sint occaecat|cupidatat non proident|sunt in culpa|qui officia|deserunt|mollit anim id est laborum"
        .components(separatedBy: "|")
    let users = ["rnystrom", "BasThomas", "jessesquires", "Sherlouk", "omwomw"]
    var autocompleteUsers = [String]()
    let tableView = UITableView()

    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.dataSource = self
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "cell")
        view.addSubview(tableView)

        borderColor = .lightGray

        messageView.textViewInset = UIEdgeInsets(top: 8, left: 8, bottom: 8, right: 16)
        messageView.font = UIFont.systemFont(ofSize: 18)

        messageView.setButton(title: "Add", for: .normal, position: .left)
        messageView.addButton(target: self, action: #selector(onLeftButton), position: .left)
        messageView.leftButtonTint = .blue
        messageView.showLeftButton = true

        messageView.setButton(inset: 10, position: .left)
        messageView.setButton(inset: 15, position: .right)

        messageView.textView.placeholderText = "New message..."
        messageView.textView.placeholderTextColor = .lightGray

        messageView.setButton(title: "Send", for: .normal, position: .right)
        messageView.addButton(target: self, action: #selector(onRightButton), position: .right)
        messageView.rightButtonTint = .blue

        messageAutocompleteController.tableView.register(UITableViewCell.self, forCellReuseIdentifier: "cell")
        messageAutocompleteController.tableView.dataSource = self
        messageAutocompleteController.tableView.delegate = self
        messageAutocompleteController.register(prefix: "@")
        
        // Set custom attributes for an autocompleted string
        let tintColor = UIColor(red: 0, green: 122/255, blue: 1, alpha: 1)
        messageAutocompleteController.registerAutocomplete(prefix: "@", attributes: [
            .font: UIFont.preferredFont(forTextStyle: .body),
            .foregroundColor: tintColor,
            .backgroundColor: tintColor.withAlphaComponent(0.1)
            ])
        
        messageAutocompleteController.delegate = self

        setup(scrollView: tableView)
    }

    @objc func onLeftButton() {
        print("Did press left button")
    }

    @objc func onRightButton() {
        data.append(messageView.text)
        messageView.text = ""
        tableView.reloadData()
        tableView.scrollToRow(
            at: IndexPath(row: data.count - 1, section: 0),
            at: .bottom,
            animated: true
        )
    }

    // MARK: UITableViewDataSource

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return tableView === self.tableView
            ? data.count
            : autocompleteUsers.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
        if tableView === self.tableView {
            cell.textLabel?.text = data[indexPath.row]
        } else {
            cell.textLabel?.text = autocompleteUsers[indexPath.row]
        }
        return cell
    }

    // MARK: UITableViewDelegate

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        if tableView === messageAutocompleteController.tableView {
            messageAutocompleteController.accept(autocomplete: autocompleteUsers[indexPath.row])
        }
    }

    // MARK: MessageAutocompleteControllerDelegate

    func didFind(controller: MessageAutocompleteController, prefix: String, word: String) {
        autocompleteUsers = users.filter { word.isEmpty || $0.lowercased().contains(word.lowercased()) }
        controller.show(true)
    }

}

